src/mlpack/methods/cf/ncf_impl.hpp

```diff
@@ -0,0 +1,440 @@
+/**
+ * @file ncf.hpp
+ * @author Haritha Nair
+ *
+ * Implementation of Neural Collaborative Filtering.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_NCF_NCF_IMPL_HPP
+#define MLPACK_METHODS_NCF_NCF_IMPL_HPP
+
+#include <mlpack/prereqs.hpp>
+#include "ncf.hpp"
+
+namespace mlpack {
+namespace cf {
+/**
+ * Construct the NCF object using the desired algorithm and optimizer.
+ */
+template<typename OptimizerType>
+NCF::NCF(arma::mat& dataset,
+         std::string algorithm,
```

Comment:
I do also believe that the algorithm type can be templatized and can be deduced during the compile time.  This can be similar to what we do with LSH, Knn, KFN. We can set the algorithm type as a policy and choose the template parameter which results in calling the corresponding constructor.